### PR TITLE
Save collaborators to report

### PR DESCRIPTION
**Pull request by jasalisbury**
**Description of change**

Includes a few additional improvements, got bigger because I refactored some tests to hopefully make them easier to deal with in the future.

 * Collaborators join table added.
 * Collaborators hooked up on the frontend. Available collaborators depends on the report's region (hardcoded to 1 for now)
 * Activity report checks permissions
 * Added the first "policies" which can be used for authorization. Should be easier to test separated from controllers
 * Route tests that hit the DB moved to the services folder, only deal with the DB interactions
 * Route tests mock out DB and authorization calls focusing on constructing correct responses based on (mocked) calls to the DB/policies
 * Multiselect sets null values to an empty array, allowing collaborators/recipients to be "zeroed" in the report
 * Approvers are scoped to region
 * The activity report now makes all API calls up front in parallel. Should keep other components simpler and cuts down on calls to the API (they can be made once instead on every form page load)
 * Added Krys and I to seeders as Admins so we should be admins in sandbox/dev until HSES resets again. (can add others to the seeders but we were the only two who currently have accounts on sandbox, so the only two I know HSES ids)
 * Updated swagger/API docs

**How to test**

1) Browse to [dev](https://tta-smarthub-dev.app.cloud.gov/activity-reports/new)
2) Fill out a report, add Chuck or Josh as a collaborator (If you aren't on the list I can add you, just ping me. Permissions should be setup correctly on Dev unless another deploy happens)
3) Verify collaborators persist across reloads
4) Send the ID of the report to someone who isn't a collaborator and verify they can't access the report

**Note**
I still need to update API docs/DB docs

**Issue(s)**
* closes https://github.com/HHS/Head-Start-TTADP/issues/109
* https://github.com/HHS/Head-Start-TTADP/issues/265
* closes https://github.com/HHS/Head-Start-TTADP/issues/271

**Checklist**

- [x] Meets issue criteria
- [x] Code tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

File: src/migrations/20210120220221-add-collaborators-to-activity-reports.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        type: Sequelize.INTEGER,
+        allowNull: false,
+        references: {
+          model: {
+            tableName: 'ActivityReports',
+          },
+          key: 'id',
+        },
+        onUpdate: 'CASCADE',
+        onDelete: 'CASCADE',
```
Comment (rahearn) [inline src/migrations/20210120220221-add-collaborators-to-activity-reports.js:20]: I always mess these up. Does this delete the ActivityReport when a collaborator is removed, or delete the collaborator automatically when an ActivityReport is deleted? (Same question for user)

File: src/models/activityReport.js [hunk 2]
```diff
@@ -16,6 +16,16 @@ export default (sequelize, DataTypes) => {
       ActivityReport.belongsTo(models.User, { foreignKey: 'lastUpdatedById', as: 'lastUpdatedBy' });
       ActivityReport.belongsTo(models.User, { foreignKey: 'approvingManagerId', as: 'approvingManager' });
       ActivityReport.hasMany(models.ActivityRecipient, { foreignKey: 'activityReportId', as: 'activityRecipients' });
+      ActivityReport.belongsToMany(models.User, {
+        through: models.ActivityReportCollaborator,
+        // The key in the join table that points to the model defined in this file
+        foreignKey: 'activityReportId',
+        // The key in the join table that points to the "target" of the belongs to many (Users in
+        // this case)
```
Comment (rahearn) [inline src/models/activityReport.js:24]: A+ comment to keep the two keys straight.

File: src/routes/activityReports/handlers.js [hunk 3]
```diff
@@ -21,13 +22,9 @@ const logContext = {
  * @param {*} res - response
  */
 export async function getApprovers(req, res) {
-  const user = await userById(req.session.userId);
-  const reportRegions = user.permissions
-    .filter((p) => p.scopeId === READ_WRITE_REPORTS)
-    .map((p) => p.regionId);
-
+  const { region } = req.query;
```
Comment (rahearn) [inline src/routes/activityReports/handlers.js:25]: Should this method should be using the Users policy? I think there are a couple of methods still with no authorization, so I'm leaning towards not making any changes here beyond the good work you've already done, and just keeping #265 open for the other routes.

File: src/routes/activityReports/handlers.test.js [hunk 4]
```diff
@@ -35,171 +24,170 @@ const mockResponse = {
...
-      permissions: [
-        {
-          userId: 51,
-          regionId: 6,
-          scopeId: SCOPES.APPROVE_REPORTS,
-        },
-      ],
+
+    it('returns the report', async () => {
+      ActivityReport.prototype.canUpdate = jest.fn().mockReturnValue(true);
```
Comment (rahearn) [inline src/routes/activityReports/handlers.test.js:52]: This is some of my javascript ignorance showing, but, how does this implementation impact the rest of the tests? Does this pollute everything? Or is it somehow scoped to this test?

**Review by rahearn — CHANGES_REQUESTED** [review]
Awesome work. I really like the policies classes.

File: src/migrations/20210120220221-add-collaborators-to-activity-reports.js (hunk 1 above)
Comment (jasalisbury) [reply src/migrations/20210120220221-add-collaborators-to-activity-reports.js:20]: I wasn't 100% sure either:

> CASCADE specifies that when a referenced row is deleted, row(s) referencing it should be automatically deleted as well.

from https://www.postgresql.org/docs/9.5/ddl-constraints.html

File: src/routes/activityReports/handlers.js (hunk 3 above)
Comment (jasalisbury) [reply src/routes/activityReports/handlers.js:25]: :+1: that sounds good to me. There may be a few other routes to update that have been added in-between starting this work and getting it merged in to main. We'll probably just want to audit all our routes on main and make sure every route added after is using a policy (or explicitly doesn't)

File: src/routes/activityReports/handlers.test.js (hunk 4 above)
Comment (jasalisbury) [reply src/routes/activityReports/handlers.test.js:52]: That is a good question, I'll look into it.

**Comment by jasalisbury** [comment]
> Awesome work. I really like the policies classes.

Stole the pattern from a great ruby library https://github.com/varvet/pundit

**Comment by rahearn** [comment]
> Stole the pattern from a great ruby library https://github.com/varvet/pundit

Probably why I like it, I use pundit on all of my ruby apps.

**Review by rahearn — APPROVED** [review]

**Pull request merged**